File: atsa/src/main/java/jp/kaiz/atsassistmod/network/payload/ControlPayloads.java
package jp.kaiz.atsassistmod.network.payload;

import jp.kaiz.atsassistmod.ATSAssistMod;
import net.minecraft.core.BlockPos;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.Identifier;

/**
 * Client → server control payloads (driver SW / ground-unit configuration).
 * Mirrors the original {@code PacketSetNotchController}, {@code PacketSetTrainState},
 * {@code PacketEmergencyBrake}, {@code PacketManualDrive}, {@code PacketTrainDriveMode},
 * {@code PacketTrainProtectionSetter}, {@code PacketGroundUnitTile(Init)}.
 */
public final class ControlPayloads {
    private ControlPayloads() {}

    private static Identifier id(String path) {
        return Identifier.fromNamespaceAndPath(ATSAssistMod.MODID, path);
    }

    // ---- driver SW (player riding the control car) ----
    public record SetNotchController(int notch) implements CustomPacketPayload {
        public static final Type<SetNotchController> TYPE = new Type<>(id("set_notch_controller"));
        public static final StreamCodec<RegistryFriendlyByteBuf, SetNotchController> CODEC =
                StreamCodec.composite(ByteBufCodecs.VAR_INT, SetNotchController::notch, SetNotchController::new);
        @Override public Type<? extends CustomPacketPayload> type() { return TYPE; }
    }

    public record SetTrainState(int stateId, int value) implements CustomPacketPayload {
        public static final Type<SetTrainState> TYPE = new Type<>(id("set_train_state"));
        public static final StreamCodec<RegistryFriendlyByteBuf, SetTrainState> CODEC = StreamCodec.composite(
                ByteBufCodecs.VAR_INT, SetTrainState::stateId,
                ByteBufCodecs.VAR_INT, SetTrainState::value,
                SetTrainState::new);
        @Override public Type<? extends CustomPacketPayload> type() { return TYPE; }
    }

    public record EmergencyBrake() implements CustomPacketPayload {
        public static final EmergencyBrake INSTANCE = new EmergencyBrake();
        public static final Type<EmergencyBrake> TYPE = new Type<>(id("emergency_brake"));
        public static final StreamCodec<RegistryFriendlyByteBuf, EmergencyBrake> CODEC =
                StreamCodec.unit(INSTANCE);
        @Override public Type<? extends CustomPacketPayload> type() { return TYPE; }
    }

    public record ManualDrive(boolean manual) implements CustomPacketPayload {
        public static final Type<ManualDrive> TYPE = new Type<>(id("manual_drive"));
        public static final StreamCodec<RegistryFriendlyByteBuf, ManualDrive> CODEC =
                StreamCodec.composite(ByteBufCodecs.BOOL, ManualDrive::manual, ManualDrive::new);
        @Override public Type<? extends CustomPacketPayload> type() { return TYPE; }
    }

    public record TrainDriveMode(int mode) implements CustomPacketPayload {
        public static final Type<TrainDriveMode> TYPE = new Type<>(id("train_drive_mode"));
        public static final StreamCodec<RegistryFriendlyByteBuf, TrainDriveMode> CODEC =
                StreamCodec.composite(ByteBufCodecs.VAR_INT, TrainDriveMode::mode, TrainDriveMode::new);
        @Override public Type<? extends CustomPacketPayload> type() { return TYPE; }
    }

    public record TrainProtectionSetter(int typeId) implements CustomPacketPayload {
        public static final Type<TrainProtectionSetter> TYPE = new Type<>(id("train_protection_setter"));
        public static final StreamCodec<RegistryFriendlyByteBuf, TrainProtectionSetter> CODEC =
                StreamCodec.composite(ByteBufCodecs.VAR_INT, TrainProtectionSetter::typeId, TrainProtectionSetter::new);
        @Override public Type<? extends CustomPacketPayload> type() { return TYPE; }
    }

    // ---- ground unit configuration ----
    /** Sets the ground-unit variant (block state TYPE), like PacketGroundUnitTileInit (id>=0). */
    public record SetGroundUnitType(BlockPos pos, int typeId) implements CustomPacketPayload {
        public static final Type<SetGroundUnitType> TYPE = new Type<>(id("set_ground_unit_type"));
        public static final StreamCodec<RegistryFriendlyByteBuf, SetGroundUnitType> CODEC = StreamCodec.composite(
                BlockPos.STREAM_CODEC, SetGroundUnitType::pos,
                ByteBufCodecs.VAR_INT, SetGroundUnitType::typeId,
                SetGroundUnitType::new);
        @Override public Type<? extends CustomPacketPayload> type() { return TYPE; }
    }

    /** Saves all editable ground-unit fields (superset of the original per-type packet). */
    public record SaveGroundUnit(BlockPos pos, boolean linkRedstone, int speed, double distance,
                                 boolean autoBrake, boolean useTrainDistance, byte[] states, int tpType)
            implements CustomPacketPayload {
        public static final Type<SaveGroundUnit> TYPE = new Type<>(id("save_ground_unit"));
        public static final StreamCodec<RegistryFriendlyByteBuf, SaveGroundUnit> CODEC = StreamCodec.of(
                (buf, p) -> {
                    buf.writeBlockPos(p.pos);
                    buf.writeBoolean(p.linkRedstone);
                    buf.writeVarInt(p.speed);
                    buf.writeDouble(p.distance);
                    buf.writeBoolean(p.autoBrake);
                    buf.writeBoolean(p.useTrainDistance);
                    buf.writeByteArray(p.states);
                    buf.writeVarInt(p.tpType);
                },
                buf -> new SaveGroundUnit(
                        buf.readBlockPos(), buf.readBoolean(), buf.readVarInt(), buf.readDouble(),
                        buf.readBoolean(), buf.readBoolean(), buf.readByteArray(), buf.readVarInt()));
        @Override public Type<? extends CustomPacketPayload> type() { return TYPE; }
    }
}

File: atsa/src/main/kotlin/jp/kaiz/atsassistmod/network/payload/ControlPayloads.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package jp.kaiz.atsassistmod.network.payload

import jp.kaiz.atsassistmod.ATSAssistMod
import net.minecraft.core.BlockPos
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.resources.Identifier

/**
 * Client -> server control payloads (driver SW / ground-unit configuration).
 * Mirrors the original packets while keeping Java record-style accessors.
 */
object ControlPayloads {
    @JvmStatic
    fun id(path: String): Identifier = Identifier.fromNamespaceAndPath(ATSAssistMod.MODID, path)

    data class SetNotchController(val notch: Int) : CustomPacketPayload {
        fun notch(): Int = notch

        override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = TYPE

        companion object {
            @JvmField
            val TYPE: CustomPacketPayload.Type<SetNotchController> = CustomPacketPayload.Type(id("set_notch_controller"))

            @JvmField
            val CODEC: StreamCodec<RegistryFriendlyByteBuf, SetNotchController> = StreamCodec.of(
                { buf, payload -> buf.writeVarInt(payload.notch) },
                { buf -> SetNotchController(buf.readVarInt()) },
            )
        }
    }

    data class SetTrainState(val stateId: Int, val value: Int) : CustomPacketPayload {
        fun stateId(): Int = stateId
        fun value(): Int = value

        override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = TYPE

        companion object {
            @JvmField
            val TYPE: CustomPacketPayload.Type<SetTrainState> = CustomPacketPayload.Type(id("set_train_state"))

            @JvmField
            val CODEC: StreamCodec<RegistryFriendlyByteBuf, SetTrainState> = StreamCodec.of(
                { buf, payload ->
                    buf.writeVarInt(payload.stateId)
                    buf.writeVarInt(payload.value)
                },
                { buf -> SetTrainState(buf.readVarInt(), buf.readVarInt()) },
            )
        }
    }

    class EmergencyBrake : CustomPacketPayload {
        override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = TYPE

        companion object {
            @JvmField
            val INSTANCE = EmergencyBrake()

            @JvmField
            val TYPE: CustomPacketPayload.Type<EmergencyBrake> = CustomPacketPayload.Type(id("emergency_brake"))

            @JvmField
            val CODEC: StreamCodec<RegistryFriendlyByteBuf, EmergencyBrake> = StreamCodec.unit(INSTANCE)
        }
    }

    data class ManualDrive(val manual: Boolean) : CustomPacketPayload {
        fun manual(): Boolean = manual

        override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = TYPE

        companion object {
            @JvmField
            val TYPE: CustomPacketPayload.Type<ManualDrive> = CustomPacketPayload.Type(id("manual_drive"))

            @JvmField
            val CODEC: StreamCodec<RegistryFriendlyByteBuf, ManualDrive> = StreamCodec.of(
                { buf, payload -> buf.writeBoolean(payload.manual) },
                { buf -> ManualDrive(buf.readBoolean()) },
            )
        }
    }

    data class TrainDriveMode(val mode: Int) : CustomPacketPayload {
        fun mode(): Int = mode

        override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = TYPE

        companion object {
            @JvmField
            val TYPE: CustomPacketPayload.Type<TrainDriveMode> = CustomPacketPayload.Type(id("train_drive_mode"))

            @JvmField
            val CODEC: StreamCodec<RegistryFriendlyByteBuf, TrainDriveMode> = StreamCodec.of(
                { buf, payload -> buf.writeVarInt(payload.mode) },
                { buf -> TrainDriveMode(buf.readVarInt()) },
            )
        }
    }

    data class TrainProtectionSetter(val typeId: Int) : CustomPacketPayload {
        fun typeId(): Int = typeId

        override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = TYPE

        companion object {
            @JvmField
            val TYPE: CustomPacketPayload.Type<TrainProtectionSetter> =
                CustomPacketPayload.Type(id("train_protection_setter"))

            @JvmField
            val CODEC: StreamCodec<RegistryFriendlyByteBuf, TrainProtectionSetter> = StreamCodec.of(
                { buf, payload -> buf.writeVarInt(payload.typeId) },
                { buf -> TrainProtectionSetter(buf.readVarInt()) },
            )
        }
    }

    /** Sets the ground-unit variant (block state TYPE), like PacketGroundUnitTileInit (id>=0). */
    data class SetGroundUnitType(val pos: BlockPos, val typeId: Int) : CustomPacketPayload {
        fun pos(): BlockPos = pos
        fun typeId(): Int = typeId

        override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = TYPE

        companion object {
            @JvmField
            val TYPE: CustomPacketPayload.Type<SetGroundUnitType> = CustomPacketPayload.Type(id("set_ground_unit_type"))

            @JvmField
            val CODEC: StreamCodec<RegistryFriendlyByteBuf, SetGroundUnitType> = StreamCodec.of(
                { buf, payload ->
                    buf.writeBlockPos(payload.pos)
                    buf.writeVarInt(payload.typeId)
                },
                { buf -> SetGroundUnitType(buf.readBlockPos(), buf.readVarInt()) },
            )
        }
    }

    /** Saves all editable ground-unit fields (superset of the original per-type packet). */
    data class SaveGroundUnit(
        val pos: BlockPos,
        val linkRedstone: Boolean,
        val speed: Int,
        val distance: Double,
        val autoBrake: Boolean,
        val useTrainDistance: Boolean,
        val states: ByteArray,
        val tpType: Int,
    ) : CustomPacketPayload {
        fun pos(): BlockPos = pos
        fun linkRedstone(): Boolean = linkRedstone
        fun speed(): Int = speed
        fun distance(): Double = distance
        fun autoBrake(): Boolean = autoBrake
        fun useTrainDistance(): Boolean = useTrainDistance
        fun states(): ByteArray = states
        fun tpType(): Int = tpType

        override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = TYPE

        companion object {
            @JvmField
            val TYPE: CustomPacketPayload.Type<SaveGroundUnit> = CustomPacketPayload.Type(id("save_ground_unit"))

            @JvmField
            val CODEC: StreamCodec<RegistryFriendlyByteBuf, SaveGroundUnit> = StreamCodec.of(
                { buf, payload ->
                    buf.writeBlockPos(payload.pos)
                    buf.writeBoolean(payload.linkRedstone)
                    buf.writeVarInt(payload.speed)
                    buf.writeDouble(payload.distance)
                    buf.writeBoolean(payload.autoBrake)
                    buf.writeBoolean(payload.useTrainDistance)
                    buf.writeByteArray(payload.states)
                    buf.writeVarInt(payload.tpType)
                },
                { buf ->
                    SaveGroundUnit(
                        buf.readBlockPos(),
                        buf.readBoolean(),
                        buf.readVarInt(),
                        buf.readDouble(),
                        buf.readBoolean(),
                        buf.readBoolean(),
                        buf.readByteArray(),
                        buf.readVarInt(),
                    )
                },
            )
        }
    }
}
